c10/util/complex.h

```diff
@@ -0,0 +1,477 @@
+  }
+
+  constexpr complex<T> &operator -=(T re) {
+    storage[0] -= re;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  constexpr complex<T> &operator *=(T re) {
+    storage[0] *= re;
+    storage[1] *= re;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  constexpr complex<T> &operator /=(T re) {
+    storage[0] /= re;
+    storage[1] /= re;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator =(const complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator +=(const complex<U> &rhs) {
+    storage[0] += rhs.real();
+    storage[1] += rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator -=(const complex<U> &rhs) {
+    storage[0] -= rhs.real();
+    storage[1] -= rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator *=(const complex<U> &rhs) {
+    // (a + bi) * (c + di) = (a*c - b*d) + (a * d + b * c) i
+    T a = storage[0];
+    T b = storage[1];
+    U c = rhs.real();
+    U d = rhs.imag();
+    storage[0] = a * c - b * d;
+    storage[1] = a * d + b * c;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator /=(const complex<U> &rhs) {
+    // (a + bi) / (c + di) = (ac + bd)/(c^2 + d^2) + (bc - ad)/(c^2 + d^2) i
+    T a = storage[0];
+    T b = storage[1];
+    U c = rhs.real();
+    U d = rhs.imag();
+    auto denominator = c * c + d * d;
+    storage[0] = (a * c + b * d) / denominator;
+    storage[1] = (b * c - a * d) / denominator;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator =(const std::complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  C10_HOST_DEVICE complex<T> &operator =(const thrust::complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+#endif
+
+  template<typename U>
+  explicit constexpr operator std::complex<U>() const {
+    return std::complex<U>(std::complex<T>(real(), imag()));
+  }
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  explicit operator thrust::complex<U>() const {
+    return thrust::complex<U>(thrust::complex<T>(real(), imag()));
+  }
+#endif
+
+  constexpr T real() const {
+    return storage[0];
+  }
+  constexpr void real(T value) {
+    storage[0] = value;
+  }
+  constexpr T imag() const {
+    return storage[1];
+  }
+  constexpr void imag(T value) {
+    storage[1] = value;
+  }
+};
+
+
+template<>
+struct alignas(8) complex<float>: public complex_common<float> {
```

Comment:
Not that it would make a difference in the current state, I think instead of hardcoding it, we should do:
```2*sizeof(float)```

Reply:
fixed